src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
                             }
                             break;
                     }
+                }
+            }
+        }
 
-                    // Determines whether node can be switched to an atomic loop.  Subsequent is the node
-                    // immediately after 'node'.
-                    static bool CanBeMadeAtomic(RegexNode node, RegexNode subsequent, int maxDepth)
-                    {
-                        if (maxDepth <= 0)
-                        {
-                            // We hit our recursion limit.  Just don't apply the optimization.
-                            return false;
-                        }
+        /// <summary>
+        /// Recurs into the last expression of a loop node, looking to see if it can find a node
+        /// that could be made atomic _assuming_ the conditions exist for it with the loop's ancestors.
+        /// </summary>
+        /// <returns>The found node that should be explored further for auto-atomicity; null if it doesn't exist.</returns>
+        private static RegexNode? FindLastExpressionInLoopForAutoAtomic(RegexNode node, int maxDepth)
+        {
+            Debug.Assert(node.Type == Loop);
 
-                        // Skip the successor down to the guaranteed next node.
-                        while (subsequent.ChildCount() > 0)
-                        {
-                            Debug.Assert(subsequent.Type != Group);
-                            switch (subsequent.Type)
-                            {
-                                case Concatenate:
-                                case Capture:
-                                case Atomic:
-                                case Require:
-                                case Loop when subsequent.M > 0:
-                                case Lazyloop when subsequent.M > 0:
-                                    subsequent = subsequent.Child(0);
-                                    continue;
-                            }
+            // Start by looking at the loop's sole child.
+            node = node.Child(0);
 
-                            break;
-                        }
+            // Skip past captures.
+            while (node.Type == Capture)
+            {
+                node = node.Child(0);
```

Comment:
this is guaranteed to never lead to a null `node` ?

Reply:
Yes, a capture will always have a child.